### ISSUE TITLE
Deprecate `ModelFilter/DatasetFilter` and remove `GeneralTags/ModelTags/DatasetTags`

### ISSUE THREAD

**Issue by Wauplin** [huggingface/huggingface_hub#1676]
From private slack thread. Listing models and datasets on the Hub with filters and tags is overkill at the moment which brings bugs (latest one [here](https://github.com/huggingface/huggingface_hub/issues/1668)) and code complexity both in `huggingface_hub` and user-side (`list_models(filter=ModelFilter(author="HuggingFaceBR4", tags=["tensorboard"]))` instead of `list_models(author="HuggingFaceBR4", filter="tensorboard")`.

2 things we can do here:
- completely drop `GeneralTags/ModelTags/DatasetTags` enums => those are >2y old and not suited anymore (requires to list ALL models from the Hub before making a selection) => already not much documented anymore
- deprecate `ModelFilter/DatasetFilter` and simplify their logic. Goal would be to pass the argument directly to list_models/list_datasets instead of an intermediate object.

This can be done in several PRs. Still has to be discussed and check impact on users (is it used in Spaces? docs? blog post? ...). Most likely no need to provide backward compatibility for enums; maybe still a bit for filters.

cc @julien-c @LysandreJik

**Comment by VictorHugoPilled** [comment]
Can I claim this?

**Comment by Wauplin** [comment]
Yes! It would be best to start by the removal of the ModelTags, DatasetTags and GeneralTags. Please let me know if you have any questions or if something is not clear enough. Thanks in advance for your contribution!

**Comment by VictorHugoPilled** [comment]
Cool, will start working today.

**Comment by druvdub** [comment]
Hi, since this issue has been left open for a bit now, I'd like to pick it up :)

**Comment by druvdub** [comment]
I would think to completely remove author from the ModelFilter as well

**Comment by Wauplin** [comment]
Hi @druvdub, first of all thanks for proposing your help on this! It is perfectly fine to take this issue yes, as it's still a topic to tackle :) 

These are good questions that I'll try to answer but first, we should distinguish two things:
1. what we aim for as a consistent API e.g. the final result
2. what we should do to reach that API without breaking existing usage

In practice, this means we will make changes twice. In a first PR we should allow the new behavior while deprecating the legacy one. In a follow-up PR (in a few months), we will completely remove the existing behavior. This transition period should allow users to adapt to the new API, or at least freeze their dependencies.

> Should I just be aiming to deprecate the classes, i.e. just add a Deprecation Warning for these filters so that people can still use them but see the warnings as well or should I be looking to completely remove them.

As said above, supporting the classes while adding deprecation warnings when used is the way to go.

> It accepts an author in filter as well as an author argument, does this not cause a conflict when looking up results?

Yep, good catch. This inconsistency is most likely due to legacy work on this part of the code. At the moment it seems that the `author` parameter is forwarded to the server as `?author={author}` while the `author` in the filter class is forwarded as `?search={author/}`. If both are set to different values, I would expect the response to be empty. What we should aim for is to support only the `?author={author}` syntax from the main `author` parameter. And since you'll be deprecating the usage of the ModelFilter class, we will get rid of the inconsistency in a few months.

> I wonder how would we want to handle if a user passes both an Iterable in filter and other filter arguments

For now you can build the query with all filters and arguments. Worse case scenario, the filters conflict with each other and the result is an empty list. In that case the user will have to adapt their query (which is fine). The raw arguments (passed to `list_models`) should keep precedence over the `ModelFilter` argument, as it's currently done.


> I would think it might be best to completely remove author from the ModelFilter as well

Not for now. In the end we aim at removing `ModelFilter` altogether. The problem with removing `author` right now is that it might be used by some users that will see their script broken just by updating the `huggingface_hub` version. This is why we need to deprecate first and remove later.

> now both trained_dataset and library can be string or a list of strings, but if trained_dataset is an empty string or an empty list it would still be added as a parameter whereas library in the same case would not be.

Good catch! Having a consistent behavior between these two arguments would definitely be better :)

---

Hope I've answered all of your questions. I'll have a look at the PR you've opened. Thanks for the contribution! :hugs:

**Comment by Wauplin** [comment]
This issue can be closed by the work from @VictorHugoPilled in https://github.com/huggingface/huggingface_hub/pull/2028 and from @druvdub in https://github.com/huggingface/huggingface_hub/pull/1761. Thanks everyone!
